source/opt/ccp_pass.cpp

```diff
@@ -0,0 +1,263 @@
+  // interesting only if the meet operation over arguments coming through
+  // executable edges yields the same constant value.
+  for (uint32_t i = 2; i < phi->NumOperands(); i += 2) {
+    if (!propagator_->IsPhiArgExecutable(phi, i)) {
+      // Ignore arguments coming through non-executable edges.
+      continue;
+    }
+    uint32_t phi_arg_id = phi->GetSingleWordOperand(i);
+    auto it = values_.find(phi_arg_id);
+    if (it != values_.end()) {
+      // We found an argument with a constant value.  Apply the meet operation
+      // with the previous arguments.
+      if (meet_val_id == 0) {
+        // This is the first argument we find.  Initialize the result to its
+        // constant value id.
+        meet_val_id = it->second;
+      } else if (it->second == meet_val_id) {
+        // The argument is the same constant value already computed. Continue
+        // looking.
+        continue;
+      } else {
+        // We found another constant value, but it is different from the
+        // previous computed meet value.  This Phi will never be constant.
+        return SSAPropagator::kVarying;
+      }
+    } else {
+      // If any argument is not a constant, the Phi produces nothing
+      // interesting for now. The propagator will callback again, if needed.
+      return SSAPropagator::kNotInteresting;
+    }
+  }
+
+  // If there are no incoming executable edges, the meet ID will still be 0. In
+  // that case, return not interesting to evaluate the Phi node again.
+  if (meet_val_id == 0) {
+    return SSAPropagator::kNotInteresting;
+  }
+
+  // All the operands have the same constant value represented by |meet_val_id|.
+  // Set the Phi's result to that value and declare it interesting.
+  values_[phi->result_id()] = meet_val_id;
+  return SSAPropagator::kInteresting;
+}
+
+SSAPropagator::PropStatus CCPPass::VisitAssignment(ir::Instruction* instr) {
+  assert(instr->result_id() != 0 &&
+         "Expecting an instruction that produces a result");
+
+  // Instructions with no input operands will always result in a varying value.
+  if (instr->NumInOperands() == 0) {
+    return SSAPropagator::kVarying;
+  }
+
+  // If this is a copy operation, and the RHS is a known constant, assign its
+  // value to the LHS.
+  if (instr->opcode() == SpvOpCopyObject) {
+    uint32_t rhs_id = instr->GetSingleWordInOperand(0);
+    auto it = values_.find(rhs_id);
+    if (it != values_.end()) {
+      values_[instr->result_id()] = it->second;
+      return SSAPropagator::kInteresting;
+    }
+    return SSAPropagator::kNotInteresting;
+  }
+
+  // Otherwise, see if the RHS of the assignment folds into a constant value.
+  std::vector<uint32_t> cst_val_ids;
+  for (uint32_t i = 0; i < instr->NumInOperands(); i++) {
+    uint32_t op_id = instr->GetSingleWordInOperand(i);
+    auto it = values_.find(op_id);
+    if (it != values_.end()) {
+      cst_val_ids.push_back(it->second);
+    } else {
+      break;
```

Comment:
How do you tell if you exited early?  Does it matter?  My guess is that this is an id that is not constant, so you do care.

Reply:
I agree with @s-perron here.  If you find an id that is non-constant, then you should bail early here rather than relying on FoldScalars later on to try to fail.  

I **could** see an instruction being foldable even with a non-constant Id.   (e.g. %mul = OpIMul %uint %x %uint_0) however, in that case how does FoldScalars line up the constant-values with the operands.  The Id that is non-constant is missing from the vector and the folding call is nonsensical.